heartbeat/monitors/active/tcp/tcp.go

```diff
@@ -39,113 +43,223 @@ func init() {
+	job, err := monitors.MakeByHostJob(
+		endpointURL.Hostname(),
+		jf.config.Mode,
+		jf.resolver,
+		monitors.MakePingIPFactory(
+			func(event *beat.Event, ip *net.IPAddr) error {
+				// use address from resolved IP
+				ipPort := net.JoinHostPort(ip.String(), endpointURL.Port())
 
-		switch scheme {
-		case "tcp", "plain", "tls", "ssl":
-		default:
-			err := fmt.Errorf("'%v' is no supported connection scheme in '%v'", scheme, h)
-			return nil, err
-		}
+				return jf.dial(event, ipPort, endpointURL)
+			}))
+	if err != nil {
+		return nil, err
+	}
+	return job, nil
+}
 
-		pair := strings.SplitN(host, ":", 2)
-		ports := config.Ports
-		if len(pair) == 2 {
-			port, err := strconv.ParseUint(pair[1], 10, 16)
-			if err != nil {
-				return nil, fmt.Errorf("'%v' is no valid port number in '%v'", pair[1], h)
-			}
+// makeSocksLookupEndpointJob makes jobs that use a Socks5 proxy to perform DNS lookups
+func (jf *jobFactory) makeSocksLookupEndpointJob(endpointURL *url.URL) (jobs.Job, error) {
+	return wrappers.WithURLField(endpointURL,
+		jobs.MakeSimpleJob(func(event *beat.Event) error {
+			hostPort := net.JoinHostPort(endpointURL.Hostname(), endpointURL.Port())
+			return jf.dial(event, hostPort, endpointURL)
+		})), nil
+}
 
-			ports = []uint16{uint16(port)}
-			host = pair[0]
-		} else if len(config.Ports) == 0 {
-			return nil, fmt.Errorf("host '%v' missing port number", h)
-		}
+// dial builds a dialer and executes the network request.
+// dialAddr is the host:port that the dialer will connect to, and where an explicit IP should go to.
+// canonicalURL is the URL used to determine if TLS is used via the scheme of the URL, and
+// also which hostname should be passed to the TLS implementation for validation of the server cert.
+func (jf *jobFactory) dial(event *beat.Event, dialAddr string, canonicalURL *url.URL) error {
+	// First, create a plain dialer that can connect directly to either hostnames or IPs
+	dc := &dialchain.DialerChain{
+		Net: dialchain.CreateNetDialer(jf.config.Timeout),
+	}
+
+	// If Socks5 is configured make that the next layer, since everything needs to go through the proxy first.
+	if jf.config.Socks5.URL != "" {
+		dc.AddLayer(dialchain.SOCKS5Layer(&jf.config.Socks5))
+	}
+
+	// Now add the IP or Hostname of the server we want to connect to.
+	// Usually this is the IP we've resolved in a prior step.
+	// If we're using a proxy with host lookup enabled the dialAddr should be the
+	// hostname we want the server to resolve for us.
+	dc.AddLayer(dialchain.ConstAddrLayer(dialAddr))
+
+	// If we're using TLS we need to add a fake layer so that the TLS layer knows the hostname we're connecting to
+	// So, the canonical URL is fixed via a ConstAddrLayer to override the TLS layer's x509 logic so it doesn't
+	// try and directly match the IP from the prior ConstAddrLayer to the cert.
+	isTLS := true
+	if canonicalURL.Scheme == "tcp" || canonicalURL.Scheme == "plain" {
+		isTLS = false
+	}
+	if isTLS {
```

Comment:
```suggestion
	if canonicalURL.Scheme != "tcp" && canonicalURL.Scheme != "plain" {
```

Reply:
yup! Committed this locally, since it's easier.